Bug 798995 - Keystrokes ignored during ledger entry

Bob-IT opened the pull request:
Some themes add additional pixels to list rows and dialogs, in this case with 'TraditionalOK' 2 pixels are added to each entry in the list and 2 to the popup. This make the calculation for the pop up height wrong and so when compared to the allocation height it is different so tries to resolve by re-popping the list and this causes an endless loop.

To fix this, use the calculated height which works for Adwaita but if this is different to the popup allocation height use that.

I have tested this on the reported OS, Fedora38 with the reported 'TraditionalOK' theme and also on an Xfce VM with the 'Clearlooks-Phenix' theme. Also built and tested on Windows and all seems to be OK.

Bob-IT committed:
Bug 798995 - Keystrokes ignored during ledger entry

Some themes add additional pixels to list rows and dialogs, in this
case with 'TraditionalOK' 2 pixels are added to each entry in the
list and 2 to the popup. This make the calculation for the pop up
height wrong and so when compared to the allocation height it is
different so tries to resolve by re-poping the list and this
cuases an endless loop.

To fix this, use the calculated height which works for Adwaita but
if this is different to the popup allocation height use that.

bparees commented:
can confirm, this appears to resolve the issue i was seeing when using `TraditionalOk` with `Mate` on `Fedora38`

thank you!

jralls commented on gnucash/register/register-gnome/completioncell-gnome.c:
```diff
@@ -926,6 +927,10 @@ popup_get_height (G_GNUC_UNUSED GtkWidget* widget,
         height = height + minh;
     }
 
+    // if popup_allocation_height set use that
+    if (box->item_edit->popup_allocation_height != -1)
+        height = box->item_edit->popup_allocation_height;
+
```
Same as above, allocation first and count * cell height as a fallback.

jralls commented on gnucash/register/register-gnome/combocell-gnome.c:
```diff
@@ -944,6 +944,10 @@ popup_get_height (G_GNUC_UNUSED GtkWidget* widget,
         height = height + minh;
     }
 
+    // if popup_allocation_height set use that
+    if (box->item_edit->popup_allocation_height != -1)
+        height = box->item_edit->popup_allocation_height;
+
```
ISTM this should go first with the `count * cell_height` as an `else` fallback. Is this partly to figure out how big a window to ask for or otherwise set `height` before the popup is realized and has an allocation?

Bob-IT commented:
I have added calculating the popup height to `gnucash-item-list` as the same calculation was being done in two places and added calling it from the `else` condition.

> Is this partly to figure out how big a window to ask for or otherwise set `height` before the popup is realized and has an allocation?

The calculation of height is correct for `Adwaita` but for other themes it may be correct or an approximation. The height is used to decide on the placement of the popup which is based on the greater space above or below the current position. Setting the height is needed when the list is large to restrict the popup to the register size.

I am not sure I can see a better way but will have a think.

ipilcher commented:
This also fixes the issue for me.  (Seen with the Breeze theme under KDE/Plasma/X11.)

jralls commented:
Thanks!
